### PR TITLE
test(autoware_test_utils): cover pure factory functions and parser nullopt branches

### PR DESCRIPTION
**Pull request by youtalk**
## Description

Implements the Recommended PR for `autoware_test_utils` from the Core quality campaign: add a unit-test suite covering the previously-untested pure factory functions and the `mock_data_parser` nullopt/empty branches, and — as part of the same change — apply the two small ABI-preserving performance/cleanup fixes the audit called out.

### Tests added

New `test/test_autoware_test_utils.cpp` (suite `AutowareTestUtils`):
- `createLaneletSegment` — id + single "lane" primitive.
- `makeCostMapMsg` — frame/info fields, exact `width*height` zero-filled data, plus a zero-size edge case.
- `makeOdometry` / `makeInitialPose` — value-checked positions and the yaw-only quaternion (`z = sin(yaw/2)`, `w = cos(yaw/2)`).
- `combineConsecutiveRouteSections` — both-non-empty (overlap drop), first-empty, second-empty branches.
- `generateTrajectory` overlap-insertion branch for both `Trajectory` and the `PathWithLaneId` specialization (size `n+1`, duplicated point).
- `resolve_pkg_share_uri` — no-match, match-and-exists, match-but-missing.

Added to `test/test_mock_data_parser.cpp` (suite `ParseFunction`):
- `parse<std::optional<LaneletRoute>>` and `parse<std::optional<PathWithLaneId>>` returning `nullopt` on missing required keys (via temp YAML files).
- `parse<std::vector<PathPointWithLaneId>>` empty-point early-return branch (default-constructed points).

### Source changes (ABI-preserving)

- `makeCostMapMsg`: replace the per-cell `push_back` loop with a single `data.assign(n_elem, 0)` (one allocation + zero-fill).
- `publishToTargetNode`: take the message by `const T &` instead of by value, avoiding a full copy of potentially large messages. The function template still accepts the same call-site arguments (lvalue or rvalue), so this is source- and ABI-compatible.

All factory signatures are unchanged; behavior is preserved.

### Verification

Built and tested in the `autoware:core-devel-jazzy` container: `colcon build` is clean and `colcon test` reports 46 tests, 0 errors, 0 failures, 9 skipped (the new gtest binary runs 21 cases, all pass; copyright/cppcheck/lint_cmake/xmllint linters pass). `pre-commit` (clang-format, cpplint, prettier, etc.) passes on all changed files.

Refs: autowarefoundation/autoware_core#1096

### CODE REVIEW

File: testing/autoware_test_utils/test/test_mock_data_parser.cpp [hunk 1]
```diff
@@ -883,4 +892,74 @@ TEST(ParseFunction, ParsePathWithLaneID)
...
+  const int fd = mkstemp(path_template);
+  if (fd != -1) {
+    ::close(fd);
+  }
+  const std::string path = path_template;
+  std::ofstream ofs(path);
+  ofs << content;
+  ofs.close();
+  return path;
+}
```
[the remark this answers is not in the document]
Comment (youtalk) [reply testing/autoware_test_utils/test/test_mock_data_parser.cpp:916]: Fixed in `1d49ede` — `write_temp_yaml` now returns an empty string when `mkstemp` fails or the file can't be opened (`!ofs`), instead of returning a non-unique/unwritable path.

File: testing/autoware_test_utils/test/test_mock_data_parser.cpp [hunk 2]
```diff
@@ -883,4 +892,74 @@ TEST(ParseFunction, ParsePathWithLaneID)
...
+  const std::string yaml = R"(
+start_pose:
+  position: {x: 1.0, y: 2.0, z: 3.0}
+  orientation: {x: 0.0, y: 0.0, z: 0.0, w: 1.0}
+segments: []
+)";
+  const auto path = write_temp_yaml(yaml);
+  const auto result = parse<std::optional<LaneletRoute>>(path);
+  EXPECT_FALSE(result.has_value());
+  std::remove(path.c_str());
```
[the remark this answers is not in the document]
Comment (youtalk) [reply testing/autoware_test_utils/test/test_mock_data_parser.cpp:933]: Fixed in `1d49ede` — added `ASSERT_FALSE(path.empty());` after `write_temp_yaml`, so the test fails clearly if the temp file can't be created/written instead of throwing inside `YAML::LoadFile`.

File: testing/autoware_test_utils/test/test_mock_data_parser.cpp [hunk 3]
```diff
@@ -883,4 +892,74 @@ TEST(ParseFunction, ParsePathWithLaneID)
...
+  const std::string yaml = R"(
+header:
+  stamp: {sec: 0, nanosec: 0}
+  frame_id: map
+points: []
+)";
+  const auto path = write_temp_yaml(yaml);
+  const auto result = parse<std::optional<PathWithLaneId>>(path);
+  EXPECT_FALSE(result.has_value());
+  std::remove(path.c_str());
```
[the remark this answers is not in the document]
Comment (youtalk) [reply testing/autoware_test_utils/test/test_mock_data_parser.cpp:950]: Fixed in `1d49ede` — added `ASSERT_FALSE(path.empty());` after `write_temp_yaml`, so the test fails clearly if the temp file can't be created/written instead of throwing inside `YAML::LoadFile`.

File: testing/autoware_test_utils/src/autoware_test_utils.cpp [hunk 4]
```diff
@@ -164,9 +164,7 @@ OccupancyGrid makeCostMapMsg(size_t width, size_t height, double resolution)
 
   // create data
   const size_t n_elem = width * height;
-  for (size_t i = 0; i < n_elem; ++i) {
-    costmap_msg.data.push_back(0.0);
-  }
+  costmap_msg.data.assign(n_elem, 0);
```
Comment (youtalk) [inline testing/autoware_test_utils/src/autoware_test_utils.cpp:167]: We need to use C++17 so far.
https://github.com/autowarefoundation/autoware_cmake/blob/main/autoware_cmake/cmake/autoware_package.cmake#L18

Comment (youtalk) [reply testing/autoware_test_utils/src/autoware_test_utils.cpp:167]: `std::vector::assign(count, value)` has been in the standard library since C++98, so this line stays within C++17 — it doesn't pull in anything newer than the `CMAKE_CXX_STANDARD 17` you linked. The C++17 differential build + the new `MakeCostMapMsg` test (which asserts `data.size() == width * height` and every cell `== 0`) are green on jazzy / humble / jazzy-agnocast:
- jazzy: https://github.com/youtalk/autoware_core/actions/runs/26765886287/job/78892060034
- humble: https://github.com/youtalk/autoware_core/actions/runs/26765886287/job/78892060003

So no standard bump is involved. If you'd still prefer the explicit `for` loop for consistency with the surrounding code, I'll revert it — just say the word.
